agent/engine/containermetadata/types.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+// dockerDummyClient is a wrapper for the docker interface functions we need
+// We use this as a dummy type to be able to pass in engine.DockerClient to
+// our functions without creating import cycles
+type dockerDummyClient interface {
+	InspectContainer(string, time.Duration) (*docker.Container, error)
+	Version() (string, error)
+}
+
+// NetworkMetadata keeps track of the data we parse from the Network Settings
+// in docker containers
+type NetworkMetadata struct {
+	networkMode string
+	ipv4Address string
+	ipv4Gateway string
+	ipv6Address string
+	ipv6Gateway string
+}
+
+// DockerContainerMD keeps track of all metadata acquired from Docker inspection
+// Has redundancies with engine.DockerContainerMetadata but this packages all
+// docker metadata we want in the service so we can change features easily
+type DockerContainerMD struct {
+	containerID   string
+	containerName string
+	imageID       string
+	imageName     string
+	ports         []api.PortBinding
+	networkInfo   NetworkMetadata
+}
+
+// TaskMetadata keeps track of all metadata associated with a task
+// provided by AWS, does not depend on the creation of the container
+type TaskMetadata struct {
+	version    string
+	clusterArn string
+	taskArn    string
+}
+
+// Metadata packages all acquired metadata and is used to format it
+// into JSON to write to the metadata file
+type Metadata struct {
```

Comment:
The struct is flattened to avoid exposing the other structs when we do JSON Marshaling in the final file. This can be done instead by a custom MarshalJSON() function but I'm not sure if this has much advantage.